Introducing dry-files (#383)
dry-rb/dry-rb.org: source/news/2021-05-04.html.markdown

---
title: Introducing dry-files
date: 2021-05-04
author: jodosha
---

We talked several times about the union of [dry-rb][dry] + [ROM][rom] + [Hanami][hanami], well today we can share good news on that front: introducing `dry-files`.

It's a gem that abstracts low level file manipulations.

> The code was originally created for `hanami-utils`, as a was to power Hanami command line.

Reviewer: Typo:

```suggestion
The code was originally created for `hanami-utils`, as a way to power Hanami command line.
```

> Then it was moved to `dry-cli`, when it was extracted by Hanami code base.

Reviewer: ```suggestion
Then it was moved to `dry-cli`, when it was extracted from the Hanami code base.
```

> Today made it finally its own debut as a standalone gem.

Reviewer: ```suggestion
Today it finally made its own debut as a standalone gem.
```

`dry-cli` is a powerful framework to build Ruby command line interfaces.

> We use it as main engine for the Hanami CLI, which also need _code generators_.

Reviewer: ```suggestion
We use it as the main engine for the Hanami CLI, which also needs _code generators_.
```

The initial idea was to have this optional `dry-cli` library to support _code generators_ via file manipulations.
But then we reached the point at which this library had more lines of code than `dry-cli` itself, so we decided to **extract** this library.

Here's a simple example:

```ruby
# frozen_string_literal: true

require "dry/files"

files = Dry::Files.new
files.write("path/to/file", "Hello, World!") # intermediate directories are created, if missing
```

`dry-files` is shipped with an extensive API to touch, (re)write, read, and remove files/directories, inject/remove/append Ruby code lines and blocks, and so on.

> Because of this abstraction we had the chance of introduce swappable adapters.

Reviewer: ```suggestion
Because of this abstraction we had the chance to introduce swappable adapters.
```
